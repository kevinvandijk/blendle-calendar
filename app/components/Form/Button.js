import React from 'react';
import { PropTypes } from '../../helpers';

const { string, oneOf } = PropTypes;

const Button = (props) => {
  let buttonClass = 'Form-Button';
  if (props.type === 'submit') buttonClass = `${buttonClass} Form-Button--isPrimary`;

  return (
    <button className={ buttonClass }>
      { props.label }
    </button>
  );
};

Button.propTypes = {
  label: string.isRequired,
  type: oneOf(['default', 'submit'])
};

Button.defaultProps = {
  type: 'default'
};

export default Button;
