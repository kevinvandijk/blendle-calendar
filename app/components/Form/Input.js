import React from 'react';
import { PropTypes } from '../../helpers';

const { oneOf, string } = PropTypes;

const Input = (props) => {
  let inputClass = 'Form-Input-field';
  if (props.type === 'date') inputClass = `${inputClass} Form-Input-date`;
  if (props.type === 'textarea') inputClass = `${inputClass} Form-Input-textarea`;

  return (
    <label htmlFor={ props.name } className="Form-Input-label">
      { props.label }

      { props.type === 'textarea'
        ? <textarea id={ props.name } className={ inputClass } />
        : <input id={ props.name } className={ inputClass } type="text" />
      }
    </label>
  );
};

Input.propTypes = {
  label: string,
  name: string.isRequired,
  type: oneOf(['text', 'textarea', 'date'])
};

Input.defaultProps = {
  type: 'text'
};

export default Input;
