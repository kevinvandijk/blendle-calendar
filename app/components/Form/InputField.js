import React from 'react';
import moment from 'moment';
import { PropTypes } from '../../helpers';

const { oneOf, string, func } = PropTypes;

class InputField extends React.Component {
  static propTypes = {
    label: string,
    name: string.isRequired,
    type: oneOf(['text', 'textarea', 'date']),
    value: string,
    onBlur: func,
    onFocus: func,
    onChange: func
  }

  static defaultProps = {
    type: 'text',
    value: ''
  }

  static contextTypes = {
    attach: func,
    detach: func
  }


  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };
  }

  state = {
    value: ''
  }

  componentWillMount() {
    if (this.context.attach) {
      this.context.attach(this);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({
        value: nextProps.value
      });
    }
  }

  componentWillUnmount() {
    if (this.context.detach) {
      this.context.detach(this);
    }
  }

  onChange = (e) => {
    let value = e.target.value;
    if (this.props.type === 'date') {
      // When entering time allow only digits and : and .
      value = value.replace(/[^\d,:,.]+/gi, '');

      // When entering time notation with : and ., the max is 5 characters
      // When only entering digits, the max is 4 characters
      if (value.match(/\d{1,2}[:,.]/g)) {
        value = value.substr(0, 5);
      } else {
        value = value.substr(0, 4);
      }
    }

    this.setState({
      value
    });

    if (this.props.onChange) this.props.onChange(value);
  }

  onBlur = () => {
    this.setState({
      focused: false
    });

    if (this.props.type === 'date' && this.state.value.length) {
      const value = this.state.value.replace(/[^\d]/g, '');

      let formattedValue;
      if (value.length <= 2) {
        formattedValue = moment().hour(value).startOf('hour').format('HH:mm');
      } else {
        const minutes = value.slice(-2);
        const hours = value.slice(0, -2);

        formattedValue = moment().hour(hours).minute(minutes).format('HH:mm');
      }

      this.setState({
        value: formattedValue
      });
    }

    if (this.props.onBlur) this.props.onBlur(this.state.value);
  }

  onFocus = () => {
    this.setState({
      focused: true
    });

    if (this.props.onFocus) this.props.onFocus();
  }

  getValue() {
    return this.state.value;
  }

  setValue(value) {
    this.setState({
      value
    });
  }

  render() {
    let inputClass = 'Form-InputField';
    const { type, name, label } = this.props;
    if (type === 'date') inputClass = `${inputClass} Form-InputField-date`;
    if (type === 'textarea') inputClass = `${inputClass} Form-InputField-textarea`;

    return (
      <label htmlFor={ name } className="Form-InputField-label">
        { label }

        { type === 'textarea'
          ?
            <textarea
              id={ name }
              className={ inputClass }
              value={ this.state.value }
              onChange={ this.onChange }
              onBlur={ this.onBlur }
              onFocus={ this.onFocus }
              ref={ (ref) => { this.input = ref; } }
            />
          :
            <input
              id={ name }
              className={ inputClass }
              type="text"
              value={ this.state.value }
              onChange={ this.onChange }
              onBlur={ this.onBlur }
              onFocus={ this.onFocus }
              ref={ (ref) => { this.input = ref; } }
            />
        }
      </label>
    );
  }
}

export default InputField;
