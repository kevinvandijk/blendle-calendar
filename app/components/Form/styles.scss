.Form-InputField {
  display: block;
  margin: 5px 0 10px;
  padding: 10px;
  width: 100%;
  border: 1px solid #DADADA;
  border-radius: 1px;
  color: darken(#DADADA, 10);
  font-size: inherit;
  font-family: inherit;
}

.Form-InputField-label {
  margin: 0;
  color: #333;
  font-size: 14px;
  line-height: 1;
}

.Form-InputField-date {
  padding-left: 38px;
  background: #fff url(../../assets/images/calendar-icon.svg) no-repeat 10px center;
}

.Form-InputField-textarea {
  height: 100px;
  resize: none;
}

.Form-Button {
  display: block;
  margin: 0 0 5px;
  padding: 14px 0;
  border: none;
  background: #DADADA;
  width: 100%;
  font-size: 12px;
  color: #fff;
  text-align: center;
  line-height: 1;
  border-radius: 3px;
  cursor: pointer;

  &:hover,
  &:active,
  &:focus {
    background: lighten(#DADADA, 5);
  }

  &.Form-Button--isPrimary {
    background: #14c674;

    &:hover,
    &:active,
    &:focus {
      background: lighten(#14c674, 5);
    }
  }

  &.Form-Button--isDisabled {
    background: lighten(#14c674, 40);
    pointer-events: none;
  }
}
