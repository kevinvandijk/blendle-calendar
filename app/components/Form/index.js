import React from 'react';
import './styles.scss';
import { PropTypes } from '../../helpers';

const { node, func } = PropTypes;

class Form extends React.Component {
  static propTypes = {
    children: node,
    onSubmit: func
  }

  static childContextTypes = {
    attach: func,
    detach: func,
    submit: func,
    reset: func
  }

  constructor(props) {
    super(props);

    this.inputs = [];
  }

  getChildContext() {
    return {
      attach: this.attach,
      detach: this.detach,
      submit: this.submit,
      reset: this.reset
    };
  }

  onKeyUp = (e) => {
    if (e.key === 'Enter') {
      const currentFocus = this.inputs.findIndex((i) => i.state.focused);
      const nextIndex = currentFocus + 1;
      if (this.inputs[nextIndex] && this.inputs[nextIndex]) {
        this.inputs[nextIndex].input.focus();
      }
    }
  }

  getValues() {
    return this.inputs.reduce((values, input) => {
      return {
        ...values,
        [input.props.name]: input.getValue()
      };
    }, {});
  }

  attach = (component) => {
    this.inputs[component.props.name] = component;
    this.inputs.push(component);
  }

  detach = (component) => {
    this.inputs = this.inputs.filter((input) => component.name !== input.name);
  }

  submit = () => {
    const values = this.getValues();

    if (this.props.onSubmit) this.props.onSubmit(values);
    this.reset();
  }

  reset = () => {
    this.inputs.forEach((input) => {
      input.setValue('');
    });
  }

  render() {
    return (
      <div onKeyUp={ this.onKeyUp }>
        { this.props.children }
      </div>
    );
  }
}

export default Form;
