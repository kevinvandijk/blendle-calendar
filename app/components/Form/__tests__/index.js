import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Form from '..';

function createMockInputComponent(name, value) {
  return {
    props: {
      name
    },
    getValue() {
      return value;
    }
  };
}

describe('Components', () => {
  describe('Form', () => {
    it('Renders itself with children', () => {
      const tree = renderer.create(
        <Form>
          <div>Form children woot</div>
        </Form>
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('Provides a childContext to attach input fields to and control the form with', () => {
      // just a sanity check so that childContext isn't accidentally deleted this would potentially
      // cause silent errors since InputFields are flexible and work without context as well
      const expected = ['attach', 'detach', 'reset', 'submit'].sort();
      expect(Object.keys(Form.childContextTypes).sort()).toEqual(expected);
    });

    it('Attaches components and is able to return their values, sorted by name', () => {
      const name1 = 'A component has no name';
      const name2 = 'A component always pays its debts'
      const value = 'Value is coming';
      const instance = shallow(<Form />).instance();

      instance.attach(createMockInputComponent(name1, value));
      instance.attach(createMockInputComponent(name2, value));

      const result = instance.getValues();

      expect(result).toEqual({
        [name1]: value,
        [name2]: value
      });
    });

    it('Detaches components and does not include their values anymore in the return value', () => {
      const mockComponent = createMockInputComponent('bla');
      const instance = shallow(<Form />).instance();

      instance.attach(mockComponent);
      instance.detach(mockComponent);

      const result = instance.getValues();

      expect(result).toEqual({});
    });

    it('Calls a given onSubmit handler on submit with the collected values', () => {
      const name = 'A component has no name';
      const value = 'Value is coming';
      const mockComponent = createMockInputComponent(name, value);
      const submitStub = jest.fn();

      const instance = shallow(<Form onSubmit={ submitStub } />).instance();
      instance.attach(mockComponent);
      instance.submit();

      expect(submitStub).toHaveBeenCalledWith({
        [name]: value
      });
    });

    it('Sets the values of attached input fields to an empty string', () => {
      const instance = shallow(<Form />).instance();
      const stub1 = jest.fn();
      const stub2 = jest.fn();
      const mock1 = createMockInputComponent('bla', 'something');
      const mock2 = createMockInputComponent('blabla', 'something else');

      mock1.setValue = stub1;
      mock2.setValue = stub2;

      instance.attach(mock1);
      instance.attach(mock2);
      instance.reset();

      expect(stub1).toHaveBeenCalledWith('');
      expect(stub2).toHaveBeenCalledWith('');
    });
  });
});
