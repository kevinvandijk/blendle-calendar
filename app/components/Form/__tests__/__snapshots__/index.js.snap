exports[`Components Form Renders itself with children 1`] = `
<div
  onKeyUp={[Function]}>
  <div>
    Form children woot
  </div>
</div>
`;
